src/main/java/net/snowflake/client/jdbc/SnowflakeBaseResultSet.java

```diff
@@ -1340,24 +1351,64 @@ public <T> T getObject(int columnIndex, Class<T> type) throws SQLException {
     logger.debug("public <T> T getObject(int columnIndex,Class<T> type)", false);
     if (SQLData.class.isAssignableFrom(type)) {
       Optional<Supplier<SQLData>> typeFactory = SnowflakeObjectTypeFactories.get(type);
-      SQLData instance =
-          typeFactory
-              .map(Supplier::get)
-              .orElseGet(() -> createUsingReflection((Class<SQLData>) type));
+      SQLData instance = (SQLData) SQLDataCreationHelper.create(type);
       SQLInput sqlInput = (SQLInput) getObject(columnIndex);
       instance.readSQL(sqlInput, null);
       return (T) instance;
+    } else if (Map.class.isAssignableFrom(type)) {
+      JsonNode jsonNode = ((JsonSqlInput) getObject(columnIndex)).getInput();
+      return (T) OBJECT_MAPPER.convertValue(jsonNode, new TypeReference<Map<String, Object>>() {});
     } else {
       return (T) getObject(columnIndex);
     }
   }
 
-  private SQLData createUsingReflection(Class<? extends SQLData> type) {
-    try {
-      return type.newInstance();
-    } catch (InstantiationException | IllegalAccessException e) {
-      throw new RuntimeException(e);
+  public <T> List<T> getList(int columnIndex, Class<T> type) throws SQLException {
+    T[] sqlInputs = getArray(columnIndex, type);
+    return Arrays.asList(sqlInputs);
+  }
+
+  public <T> T[] getArray(int columnIndex, Class<T> type) throws SQLException {
+    Optional<Supplier<SQLData>> typeFactory = SnowflakeObjectTypeFactories.get(type);
+    Map[] jsonMaps = (Map[]) getArray(columnIndex).getArray();
+    T[] arr = (T[]) java.lang.reflect.Array.newInstance(type, jsonMaps.length);
+    int counter = 0;
+    for (Map map : jsonMaps) {
+      SQLData instance = (SQLData) SQLDataCreationHelper.create(type);
+      SQLInput sqlInput =
+          new JsonSqlInput(
+              OBJECT_MAPPER.convertValue(map, JsonNode.class),
+              session,
+              sfBaseResultSet.getConverters(),
+              sfBaseResultSet.getMetaData().getColumnMetadata().get(columnIndex - 1).getFields());
+      instance.readSQL(sqlInput, null);
+      arr[counter++] = (T) instance;
+    }
+
+    return arr;
+  }
+
+  public <T> Map<String, T> getMap(int columnIndex, Class<T> type) throws SQLException {
+    Optional<Supplier<SQLData>> typeFactory = SnowflakeObjectTypeFactories.get(type);
+    Object object = getObject(columnIndex);
+    JsonNode jsonNode = ((JsonSqlInput) object).getInput();
+    Map<String, Object> map =
+        OBJECT_MAPPER.convertValue(jsonNode, new TypeReference<Map<String, Object>>() {});
+    Map<String, T> resultMap = new HashMap<>();
+
+    for (Map.Entry<String, Object> entry : map.entrySet()) {
+      SQLData instance = (SQLData) SQLDataCreationHelper.create(type);
+      SQLInput sqlInput =
```

Comment:
This code (creating SQLData) is copied multiple times. Maybe we can introduce a function?

Reply:
It's one-liner now and extracted to external class. For me it isn't candidate to be a function